### ISSUE TITLE
AUDIT | Customise CreatedBy

### ISSUE THREAD

**Issue by KarishImrith** [zzzprojects/EntityFramework-Plus#64]
Hi,

My goal is to change the value of the CreatedBy field. When attempting to do this as per your wiki documentation, the value stored did not change. I even tried using your CreatedDate example as it is and still no change to what is stored in my AuditEntry table.

**DbContextConstructor:**

``` csharp
public ApplicationDbContext() : base("DefaultConnection")
{
    //Customise CreatedDate
    AuditManager.DefaultConfiguration
    .AuditEntryFactory = args =>
        new AuditEntry() { CreatedDate = DateTime.Now.AddHours(-2) };

    //Enable auto audit
    AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
        (context as ApplicationDbContext).AuditEntries.AddRange(audit.Entries);

}
```

**SaveChanges overridden:**

``` csharp
public override int SaveChanges()
{
    var audit = new Audit();
    audit.PreSaveChanges(this);
    var rowAffecteds = base.SaveChanges();
    audit.PostSaveChanges();

    if (audit.Configuration.AutoSavePreAction != null)
    {
        audit.Configuration.AutoSavePreAction(this, audit);
        base.SaveChanges();
    }

    return rowAffecteds;
}

public override Task<int> SaveChangesAsync()
{
    return SaveChangesAsync(CancellationToken.None);
}

public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken)
{
    var audit = new Audit();
    audit.PreSaveChanges(this);
    var rowAffecteds = await base.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
    audit.PostSaveChanges();

    if (audit.Configuration.AutoSavePreAction != null)
    {
        audit.Configuration.AutoSavePreAction(this, audit);
        await base.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
    }

    return rowAffecteds;
}
```

**DbSets:**

``` csharp
public DbSet<AuditEntry> AuditEntries { get; set; }
public DbSet<AuditEntryProperty> AuditEntryProperties { get; set; }
```

**Documentation I am using:**

``` csharp
// Customize CreatedDate

// using Z.EntityFramework.Plus; // Don't forget to include this.

// Globally
AuditManager.DefaultConfiguration
            .AuditEntryFactory = args => 
                new AuditEntry() { CreatedDate = DateTime.Now.AddHours(-5) };

// By Instance
var audit = new Audit();
audit.CreatedBy = "ZZZ Projects"; // Optional
audit.Configuration.AuditEntryFactory = args => 
                       new AuditEntry() {CreatedDate = DateTime.Now.AddHours(-5)};
ctx.SaveChanges(audit);
```

Edit Post:

``` csharp
[HttpPost]
[ValidateAntiForgeryToken]
public ActionResult Edit(Customer customer)
{
    if (ModelState.IsValid)
    {
        db.Customers.AddOrUpdate(customer);
        if (db.Entry(customer).State == EntityState.Added) throw new Exception("The data processed has not been picked up as a modification.");
        db.SaveChanges();

        return RedirectToAction("Index");
    }
    return View(customer);
}
```

The CreatedDate value is not changing as per my customization for all operations (add, modify, etc).
In the interim I am using the following to get my results:

``` csharp
public ApplicationDbContext() : base("DefaultConnection")
{
    //Customise CreatedDate
    //AuditManager.DefaultConfiguration
    //.AuditEntryFactory = args =>
    //    new AuditEntry() { CreatedDate = DateTime.Now.AddHours(-2) };

    //Enable auto audit
    AuditManager.DefaultConfiguration.AutoSavePreAction = (context, audit) =>
        (context as ApplicationDbContext).AuditEntries.AddRange(audit.Entries.Select(c => { c.CreatedBy = "Hi"; return c; }));

}
```

Please advise.

Warm Regards

**Comment by KarishImrith** [comment]
Apologies for the format of the post - I am new to this :)

**Comment by zzzprojects** [comment]
don't worry ;) here is how you use code blocks:
https://help.github.com/articles/creating-and-highlighting-code-blocks/

**Comment by zzzprojects** [comment]
You are right,

We successfully reproduced the issue. It seems we forget to add this property in our Clone method

A new version should be available in a few hour with this fix,

Best Regards,

Jonathan

**Comment by zzzprojects** [comment]
Hello @KarishImrith ,

Sorry for the delay.

The v1.4.9 is now available with this fix.

Let us know if the CreatedDate is now working.

Best Regards,

Jonathan

**Comment by zzzprojects** [comment]
Hello @KarishImrith ,

We will close this issue since we believe we fixed it in v1.4.9.

Just re-open it if you still have the issue.

Best Regards,

Jonathan
